File: downloadablebookandexercisecode/1256/code_files/Chapter_03_Complexity_Analysis/CSES_sum_of_two_values.py
# https://cses.fi/problemset/task/1640
import copy

n, x = map(int, input().strip().split())
a = list(map(int, input().strip().split()))

# we can use python methods to get the index as well,
# but i will do it very simply here.
# among the other options:
# using numpy argsort
# zip with index
# sorted

a_orig = copy.deepcopy(a)

a.sort()

i0 = 0
i1 = n-1
found = False

while i0 < i1:
    if a[i0] + a[i1] > x:
        i1 -= 1
        continue
    if a[i0] + a[i1] == x:
        found = True
        break
    if a[i0] + a[i1] < x:
        i0 += 1
        continue

if not found:
    print("IMPOSSIBLE")
else:
    id1 = a_orig.index(a[i0])
    id2 = a_orig.index(a[i1])
    if id2 == id1:
        id2 = a_orig.index(a[i1], id1 + 1)
    print(f'{id1+1} {id2+1}')

File: downloadablebookandexercisecode/1256/code_files/Chapter_03_Complexity_Analysis/CSES_sum_of_two_values.cpp
// https://cses.fi/problemset/task/1640

#include <iostream>
#include <algorithm>

using namespace std;

int main() {

    int n, x;
    cin >> n >> x;

    long long a[n]; // need to use long long
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
    }

    long long b[n];
    copy(a, a + n, b);

    sort(a, a + n);
    int i0 = 0;
    int i1 = n - 1;
    bool found = false;
    while (i0 < i1) {
        if (a[i0] + a[i1] > x) {
            i1--;
            continue;
        }
        if (a[i0] + a[i1] == x) {
            found = true;
            break;
        }
        if (a[i0] + a[i1] < x) {
            i0++;
            continue;
        }
    }

    if (!found) {
        cout << "IMPOSSIBLE";
        exit(0);
    }

    int val1 = a[i0], id1 = -1;
    int val2 = a[i1], id2 = -1;
    for (int i = 0; i < n; ++i) {
        if (id1 == -1 && b[i] == val1) id1 = i;
        if (id2 == -1 && id1 != i && b[i] == val2) id2 = i;
    }
    cout << id1 + 1 << " " << id2 + 1;
    
}
